### DISCUSSION TITLE
Enable turbo on non-monorepo projects

### DISCUSSION THREAD

**Discussion by NateRadebaugh** [vercel/turborepo#376] [Ideas]
### Describe the feature you'd like to request

It seems like the only supported scenario is to use turbo on monorepos. I'd imagine similar perf improvements through smart granular caching could be had on non-repo projects as well at the file-level, etc.

Feature request: enable turbo on non-monorepo projects

### Describe the solution you'd like

Documented `turbo` configuration should work without any `workspaces` configured.

### Describe alternatives you've considered

The alternative would be to put my Next.js website into a separate `apps/web` dir and configure it as a monorepo with a single app, but this feels like overkill and doesn't work "out of the box" with vercel CI builds.

**Comment by jaredpalmer** [comment]
@rauchg and I were just taking about this the other day! It’s a great idea

**Comment by xcv58** [comment]
+1 for the idea to support non-monorepo projects.

And there is another use case similar to this: support the scripts in the top-level `package.json` within a yarn workspace.

For example, some projects have top-level `lint` and `start` commands.

**Comment by jaredpalmer** [comment]
Yep! Top-level commands are a must-have

**Comment by boennemann** [comment]
I thought about this, too, and I basically miss two things, which turn out to be pretty much the same. (Because I think the configuration for those two could basically be the same)

1) work on non-monorepo projects
2) have package-level override

Imagine a ts app/project app that needs to generate code before it works, e.g. https://www.graphql-code-generator.com/ or https://www.prisma.io/.

Yes, it's definitely possible to pull these things out into their own package, but I think a separate package should be used for logical units and not be a work-around to enable caching.

Especially with graphql files I really like the colocation with the implementation, so it's a real pain to move all of them to a whole different folder/package just to have the code generation cached.

I think next to "pipelines", there could be a separate "package" config entry. This config is either taken into account for single package.json non-monorepo projects, or within the monorepo on the individual package's package.json. The parent turbo could then look at these individual overrides.


Codegen Example:

```json
{
  "turbo": {
    "package": {
      "build": {
        "dependsOn": [
          "codegen"
        ]
      },
      "codegen": {
        "inputs": [
          "src/graphql/**.*graphql"
        ],
        "outputs": [
          "schema.graphql",
          "src/graphql/generated.ts"
        ]
      }
    }
  }
}
```

Prisma Example
```json
{
  "turbo": {
    "package": {
      "build": {
        "dependsOn": [
          "prisma:generate"
        ]
      },
      "prisma:generate": {
        "inputs": [
          "prisma/prisma.schema"
        ],
        "outputs": [
          "node_modules/.prisma/client"
        ]
      }
    }
  }
}
```

**Comment by dvakatsiienko** [comment]
Is this in progress? Currently remote and local caches gets diverged in non-monorepo project, even a basic one one from [non-monorepo](https://github.com/vercel/turbo/tree/main/examples/non-monorepo) example.

**Comment by anthonyshew** [comment]
Turborepo now supports non-monorepos today!
